### PR TITLE
in progress pagination code

### PR DESCRIPTION
**Pull request by manni8436**
…into manni# Hackathon Git Labs Project PR 🎉🎉

👉 _**Please remove the below and replace with your own values, leaving the headers where they are.**_ 👈

## PR contains:
- Created function for Pagination.
- Declared some variables outside Pagination function.
- Added arguments to `createParticipantesCards`.
- Refactored pagination code.

## Hacktoberfest 2022:
- [x] Does this PR follow the `intermediate.md` guide and submitted as part of Hacktoberfest 2022?

_Please note, we cannot accept `basic.md` PR submissions for Hacktoberfest, it must be a custom HTML or CSS creation following the `intermediate.md` guide._

## Breaking changes
None

## Screenshots
image

image

image

### CODE REVIEW

**Review by andrewdempsey2018 — COMMENTED** [review]
This looks great so far @manni8436 keep up the good work!

**Comment by auxfuse** [comment]
Hi @manni8436 thanks for the PR man 👏 

Because the JS has gone through some iterations with @seanyoung247 over the last few weeks, I'm going to tag him in here _gently_ 😅 and respectfully ask for his own review of this and to see if we can resolve the Merge Conflict without too much jostling of the current work, and the work done in this PR?

**Comment by manni8436** [comment]
Hi @auxfuse, it was a great experience creating the pagination, had a few tough points that took a little while to overcome, but it was a great experience, I've learnt a lot, help me understand JS more, try new things and show that my JS logic and understanding has improved over the last month or so 😄 🤓

**Comment by seanyoung247** [comment]
I did somewhat massively refactor main.js, yes 😱...

The card creation functions have moved to a separate module, so it might be a good idea to do a pull to get the moved code. For the most part pulling the change to the code layout should solve a chunk of the conflicts.

My only other thought for resolving conflicts is that adding pagination shouldn't need to alter `createParticipantesCards`. It's purpose is just to create the cards asked for, it doesn't care what page they're on. `currentPage` and `perPage` don't appear to be used in the function, so is that change to the function signature necessary?

**Review by seanyoung247 — CHANGES_REQUESTED** [review]
Looks good, shouldn't need much to resolve conflicts.

File: assets/js/main.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ const generateCardFromTemplate = (participant, emojis, template) => {
  * @param {Array} indexes - (Optional) List of participant indexes to use. If not present all participants are created in order. Indexes are assumed to be valid.
  * @param {Boolean} pages - (Optional) If true the partipant will only be added if they have a custom page
  */
-const createParticipantesCards = (participants, emojis, elements, indexes = null, pages = false) => {
+const createParticipantesCards = (participants, emojis, elements, currentPage, perPage, indexes = null, pages = false) => {
```
[the remark this answers is not in the document]
Comment (manni8436) [reply assets/js/main.js:59]: No, we don't need `currentPage and perPage` as arguments from anywhere, but the only one I need is `indexes=pageParticipants`

File: assets/js/main.js [hunk 2]
```diff
@@ -131,17 +131,80 @@ const createSkeletonLoaders = elements => {
...
 
   });
 
-})();
\ No newline at end of file
+})();
+
+// pagination
+
+const paginationEl = document.getElementById("pagination");
```
[the remark this answers is not in the document]
Comment (manni8436) [reply assets/js/main.js:155]: Yes, it should be fine as a local scope.

File: assets/js/main.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (seanyoung247) [reply assets/js/main.js:59]: I think removing currentPage and perPage and pulling the current state should resolve most if not all the merge conflicts then.

[the remark this answers is not in the document]
Comment (manni8436) [reply assets/js/main.js:59]: Cool, do what you need to do 😄

[the remark this answers is not in the document]
Comment (seanyoung247) [reply assets/js/main.js:59]: That'll have to come from your end I'm afraid, I don't have write privileges on your branch ;). If you do a pull from upstream it should drag down the changes to main.js, then just remove the extra parameters.

[the remark this answers is not in the document]
Comment (manni8436) [reply assets/js/main.js:59]: Ah yes, fair enough, I will get those changes done send it over again 😄

**Comment by manni8436** [comment]
I've made those changes and pushed the code again, if someone can have a look now please 😊

**Comment by seanyoung247** [comment]
@manni8436 Have you also pulled from upstream?

**Comment by manni8436** [comment]
Yes I did but when I do that I seem to be getting this issue

image

**Comment by auxfuse** [comment]
Hi @manni8436 I've resolved that pathing error. Can you do another `git pull upstream main` please, and then push your changes back to the branch associated with this PR? At the moment you are still showing as having merge conflicts in this PR. And the error you are receiving in your terminal has nothing to do with the merge conflict.

Although, I do admit, you will most likely get an extra merge conflict now because I have fixed that pathing error 😆 And in that case, you are free to just accept the upstream changes to resolve. 🤜

**Comment by auxfuse** [comment]
For additional reference, this is where we are still the original merge conflict present. 👇 👇 
image

**Comment by seanyoung247** [comment]
> For additional reference, this is where we are still the original merge conflict present. point_down point_down image
I would suggest the resolution would be:
```
communityElements.container.innerHTML = "";
createParticipantesCards(participants, emojis, communityElements, indexes=pageParticipants);
```

**Comment by manni8436** [comment]
Thanks @auxfuse I'll do it shortly and let you know when it's done 😊

File: assets/js/main.js [hunk 3]
```diff
@@ -58,13 +58,14 @@ document.getElementById('showcases-slideshow')
       // Alter count to show all cards
       communityElements.count = participants.length;
       createParticipantesCards(participants, emojis, communityElements, indexes=pageParticipants);
+      communityElements.container.innerHTML = "";
+      createParticipantesCards(participants, emojis, communityElements);
```
Comment (seanyoung247) [inline assets/js/main.js:62]: Currently this is deleting the paginated cards added on line 60, then adding cards for all participants.

These lines:
```
      createParticipantesCards(participants, emojis, communityElements, indexes=pageParticipants);
      communityElements.container.innerHTML = "";
      createParticipantesCards(participants, emojis, communityElements);
```
Probably need to be:
```
      communityElements.container.innerHTML = "";
      createParticipantesCards(participants, emojis, communityElements, indexes=pageParticipants);
```

**Review by seanyoung247 — CHANGES_REQUESTED** [review]

File: assets/js/main.js [hunk 4]
```diff
@@ -53,10 +57,67 @@ document.getElementById('showcases-slideshow')
...
+  
+  let firstParticipantIndex = currentPage * perPage -6;
+  let lastParticipantIndex = currentPage * perPage -1;
+  let pageParticipants = [];
+  
+  if (lastParticipantIndex >= participants.length) {
+    lastParticipantIndex = participants.length -1;
+  };
+  
+  for (let i = firstParticipantIndex; i <= lastParticipantIndex; i++) {
```
Comment (seanyoung247) [inline assets/js/main.js:119]: Minor point, but this is doing the same job as the utility function genIndexes, so could be simplified to:
`return genIndexes(lastParticipantIndex+1, firstParticipantIndex);`

File: assets/js/main.js [hunk 5]
```diff
@@ -53,10 +57,67 @@ document.getElementById('showcases-slideshow')
...
+  paginationPrevButton.textContent = "Previous";
+  paginationPrevButton.href = `?page=${currentPage - 1}`;
+  paginationEl.appendChild(paginationPrevLi);
+  paginationPrevLi.append(paginationPrevButton);
+
+  if (currentPage === 1) {
+    paginationPrevLi.style.display = "none";
+  };
+  
+  let firstParticipantIndex = currentPage * perPage -6;
```
Comment (seanyoung247) [inline assets/js/main.js:111]: Should this be:
`  let firstParticipantIndex = currentPage * perPage - perPage;`?

**Review by seanyoung247 — CHANGES_REQUESTED** [review]

File: assets/js/main.js (hunk 3 above)
Comment (manni8436) [reply assets/js/main.js:62]: OK, I've made those changes

File: assets/js/main.js [hunk 6]
```diff
@@ -53,10 +57,67 @@ document.getElementById('showcases-slideshow')
...
+  let pageParticipants = [];
+  
+  if (lastParticipantIndex >= participants.length) {
+    lastParticipantIndex = participants.length -1;
+  };
+  
+  for (let i = firstParticipantIndex; i <= lastParticipantIndex; i++) {
+    pageParticipants.push(i);
+  }
+  return pageParticipants;
```
Comment (seanyoung247) [inline assets/js/main.js:122]:
```
  for (let i = firstParticipantIndex; i <= lastParticipantIndex; i++) {
    pageParticipants.push(i);
  }
  return pageParticipants;
```
Minor point, but this is doing the same job as the utility function genIndexes, so could be simplified to:
`return genIndexes(lastParticipantIndex+1, firstParticipantIndex);`

File: assets/js/main.js (hunk 5 above)
Comment (manni8436) [reply assets/js/main.js:111]: I guess it can be, since it does the same exact thing 🤔, I've made those changes

File: assets/js/main.js (hunk 6 above)
Comment (manni8436) [reply assets/js/main.js:122]: I've also made these changes now

**Comment by manni8436** [comment]
Right, I think I have it done now and up to date, and it seems to be working the way it should. 

I've made those changes that have been mentioned above and going to push them through now.

Unless there are some things that need to be changed??

**Comment by seanyoung247** [comment]
@manni8436  The test deployment is throwing an exception, does it work as expected your end?

**Comment by seanyoung247** [comment]
I've found the problem. What happens if the page url parameter is missing or has an invalid value? Say 0 or 1000?

**Review by seanyoung247 — CHANGES_REQUESTED** [review]
Needs input validation on `currentPage`. What happens if the page parameter is missing or an invalid value?

We should ensure `currentPage` is clamped to the range 1 to max pages.

**Comment by manni8436** [comment]
Actually I have noticed that issue and I have yet to fix that, what's the best way to cap that between the first page and the last page??

**Review by andrewdempsey2018 — COMMENTED** [review]
Conflicts resolved thanks @seanyoung247 I'll leave this open for now as I can see @manni8436 is still working away

**Comment by seanyoung247** [comment]
https://github.com/auxfuse/hackathon-git-labs/pull/168#issuecomment-1288333451
Could be as simple as a clamp function on currentPage just after you've calculated Mac pages.

**Comment by auxfuse** [comment]
Was the clamp function put in place before this was merged lads? @andrewdempsey2018 @seanyoung247 @manni8436 are we happy that this is done and we can implement more for future enhancements?

**Comment by seanyoung247** [comment]
@seanyoung247 It doesn't appear to be in the current code.

**Pull request merged**